feat: Enum improvements

magicus opened the pull request:
This PR contains two related but yet sort of different changes. Sorry, but it was too hard to untangle them.

1) All enums are renamed to use the UPPER_CAMEL standard.

2) Our "import/export" of enums have been improved. That is, when reading and writing config files. in the `/config` command and in the settings screen. New behavior:

* In the json config file, and the `/config` command, the enums will be converted to lowerCamel, to better fit in with the identifiers in the surrounding context.
* When printing the value of an enum with `/config`, or in the toggle-button in the config GUI, a "nice" name is shown instead. This is the value of `toString()` if it is overridden (we have that for some enums, but currently not for any enum that is a Config), or the UPPER_CAMEL name is automatically translated to Title Case.
* Enums (and booleans) now get tab completion for `/config set`.

An Upfixer has been written to transition UPPER_CAMEL config to lowerCamel. Some changes to the Upfixer framework were necessary. I also made some not strictly necessary changes, wherein the upfixer name is derived automatically (and so that too is lowerCamel, which fits better in the config file). That means that existing upfixers will run again, since they now have a new name. I inspected the code and found that to be okay for all but one, which has therefore kept its old name.

magicus committed:
refactor: Rename all enum to follow the UPPER_CAMEL name rule

magicus committed:
fix: The config UI reset button has stopped working

magicus committed:
* Change json for enum to use lowerCamel
* Introduce Upfixer to convert enum format
* Modify Upfixer framework to support use case
* Improve upfixer naming
* Use nice name of enums in settings UI

magicus committed:
Support enums nicely with /config command

magicus committed:
feat: More robust handling of null in config file

If the config file accidentally contains a null value, it will now be replaced by the default value, instead of causing a crash later on.

magicus committed:
Use a TypeAdapterFactory instead, to reach embedded enums

magicus committed:
Merge branch 'main' into disallow-config-null

magicus committed:
Merge branch 'disallow-config-null' into enum-name-format

magicus committed:
feat: Remove @TypeOverride

Replace this by proper detection of generic types. This cannot be done in runtime on an object due to type erasure, but it can be extracted from the declaring field, which we actually check.

Also wrap the type so declared abstract collection types are replaced by concrete implementations suitable for json serialization.

magicus committed:
Merge branch 'remove-type-override' into enum-name-format

magicus committed:
Fix merge error

magicus committed:
Remove allowNull annotation

magicus committed:
Qodana is right...

magicus committed:
Fix enum upfixer to work with embedded enums as well

magicus committed:
JsonElement.equals is stupid

magicus committed:
How did that end up committed? :-o

magicus committed:
Get rid of getClassOfConfigField()

magicus committed:
Merge branch 'main' into enum-name-format

magicus commented:
Note that this depends on #1399, which is incorporated in this PR. It might be easier to review this once that has been integrated.

magicus commented:
I realize this is a big PR, but it touches several interrelated things and I find it hard to see a way how to split this in several PRs. Much of the "meat" of the PR is concentrated to a few files, and most changes are just enum renames that can be trivially skipped over.

kristofbolyai committed:
Merge branch 'main' into enum-name-format

kristofbolyai reviewed (approved):
Looks good, I could not find anything strange.

magicus committed:
feat: Make reset button grayed-out if no change has been made (#1406)

This makes it easy to spot if a setting is left at default or has been changed by the user

magicus committed:
ci: spotless formatting

magicus committed:
Merge branch 'main' into enum-name-format

magicus committed:
Fix merge errors
